Rely on the new modules loading logic instead of a custom hook

nicolo-ribaudo opened the pull request:
Hi!

This PR includes the spec for import reflection aligned to what I believe is the shared vision we built during the last few months.

[**PREVIEW HERE**](https://nicolo-ribaudo.github.io/proposal-import-reflection/)

- `import module mod from "x"` and `await import("x", { module: true })` give you a `Module` object as defined in Layer 0 of the compartments proposal ([spec](https://tc39.es/proposal-compartments/0-module-and-module-source.html), [explainer](https://github.com/tc39/proposal-compartments/blob/master/0-module-and-module-source.md)).
   - the logic for static imports is in all the changes to the [ECMAScript Language: Scripts and Modules](https://nicolo-ribaudo.github.io/proposal-import-reflection/#sec-ecmascript-language-scripts-and-modules) section;
   - the logic for dynamic imports is in step 6 of [`EvaluateImportCall`](https://nicolo-ribaudo.github.io/proposal-import-reflection/#sec-evaluate-import-call) (to parse the "options" object) and step 4 of [`ContinueDynamicImport`](https://nicolo-ribaudo.github.io/proposal-import-reflection/#sec-ContinueDynamicImport) (to return the `Module` object).
- `import(mod)`, where `mod instanceof Module`, links and evaluates the module. This behavior is identically defined in Layer 0 of compartments and in the [Module Blocks](github.com/tc39/proposal-js-module-blocks) proposal, but this duplication is necessary to make each proposal usable on its own.
   - this is specified in step 8 of [`EvaluateImportCall`](https://nicolo-ribaudo.github.io/proposal-import-reflection/#sec-evaluate-import-call).
- `Module.prototype.source` gives you "the source object of the module instance". This proposal doesn't define _what_ it is, it only define that it is an object or null. The compartments proposal ([layer 0](https://github.com/tc39/proposal-compartments/blob/master/0-module-and-module-source.md) and [layer 2](https://github.com/tc39/proposal-compartments/blob/master/2-virtual-module-source.md)) better define what is is, but tl;dr it's morally a context-free unlinked representation of the module. This proposal needs it so that `import module x from "x.wasm"; x.source` can give access to the raw `WebAssembly.Module` object.

Also, this PR is written on top of https://github.com/tc39/ecma262/pull/2905. The `ecma262-biblio.json` file you see is to ensure proper linking of all the new AOs to that PR, until it's merged.

cc @littledan, @kriskowal and @caridy, who work on other pieces of this modules improvements effort.

ljharb commented:
Does your OP encapsulate the entirety of the proposal? Meaning, there's no other semantics than what you've described?

nicolo-ribaudo commented:
Yes. Or at least, it encapsulates the entirety of one of the possible versions of this proposal: the one which leaves less customizations to the host.

littledan commented:
An essential aspect of this proposal will be the Wasm+HTML integration specification, where the .source of a Wasm module is the WebAssembly.Module. I suspect that we could do this without requiring the other compartments proposals to complete, just by saying it is undefined for JS modules for now, and returns that particular object for WebAssembly.

nicolo-ribaudo commented:
Right now this PR makes it `null` for JS modules, and mentions `WebAssembly.Module` in the note of https://nicolo-ribaudo.github.io/proposal-import-reflection/#sec-module.prototype.source.

guybedford commented on spec.emu:
```diff
@@ -6,77 +6,178 @@
-            1. IfAbruptRejectPromise(_reflection_, _promiseCapability_).
-            1. If _reflection_ is not *undefined*,
-              1. If _reflection_ is not `"module"`, then
+            1. Let _moduleOption_ be Get(_options_, *"module"*).
```
```suggestion
            1. Let _reflectOption_ be Get(_options_, *"reflect"*).
```

guybedford replied:
The current specification defines a `reflect` option here.

guybedford commented on spec.emu:
```diff
@@ -6,77 +6,178 @@
-            1. If _reflection_ is not *undefined*,
-              1. If _reflection_ is not `"module"`, then
+            1. Let _moduleOption_ be Get(_options_, *"module"*).
+            1. IfAbruptRejectPromise(_moduleOption_, _promiseCapability_).
+            1. If _moduleOption_ is not *undefined*,
+              1. If Type(_moduleOption_) is not Boolean, then
```
```suggestion
            1. IfAbruptRejectPromise(_reflectOption_, _promiseCapability_).
            1. If _reflectOption_ is not *undefined*,
              1. If Type(_reflectOption_) is not *"module"*, then
```

guybedford commented on spec.emu:
```diff
@@ -6,77 +6,178 @@
                 1. Perform ! Call(_promiseCapability_.[[Reject]], *undefined*, &laquo; a newly created *TypeError* object &raquo;).
                 1. Return _promiseCapability_.[[Promise]].
-          1. Perform HostImportModuleDynamically(_referencingScriptOrModule_, _specifierString_, _reflection_, _promiseCapability_).
+              1. Set _importAsModuleObject_ to _moduleOption_.
```
```suggestion
              1. Set _importReflection_ to ~module~.
```

guybedford commented on spec.emu:
```diff
@@ -6,77 +6,178 @@
                 1. Return _promiseCapability_.[[Promise]].
-          1. Perform HostImportModuleDynamically(_referencingScriptOrModule_, _specifierString_, _reflection_, _promiseCapability_).
+              1. Set _importAsModuleObject_ to _moduleOption_.
+          1. Let _state_ be the DynamicImportState Record { [[PromiseCapability]]: _promiseCapability_, [[ImportAsModuleObject]]: _importAsModuleObject_ }.
```
```suggestion
          1. Let _state_ be the DynamicImportState Record { [[PromiseCapability]]: _promiseCapability_, [[ImportReflection]]: _importReflection_ }.
```

guybedford commented on spec.emu:
```diff
@@ -605,8 +844,90 @@ contributors: Luca Casonato, Guy Bedford
+  <h1>Fundamental Objects</h1>
+
+  <emu-clause id="sec-module-objects" number="6">
+    <h1><ins>Module Objects</ins></h1>
```
We shouldn't need to define any of the `Module` object at this stage if we can null out the JS case.

guybedford commented on spec.emu:
```diff
@@ -6,77 +6,178 @@
+          1. If Type(_specifier_) is not Object or _specifier_ does not have a [[ModuleRecord]] internal slot, then
+            1. Set _specifier_ to Completion(ToString(_specifier_)).
+            1. IfAbruptRejectPromise(_specifier_, _promiseCapability_).
+          1. Let _importAsModuleObject_ be *false*.
```
This should be renamed to import reflection:

```suggestion
          1. Let _importReflection_ be *null*.
```

guybedford commented on spec.emu:
```diff
@@ -6,77 +6,178 @@
+            1. Return ~unused~.
+          1. Let _module_ be _moduleCompletion_.[[Value]].
+          1. <ins>If _state_.[[ImportAsModuleObject]] is *true*, then</ins>
+            1. <ins>Let _moduleObject_ be GetModuleObject(_module_).</ins>
```
Rather than `GetModuleObject`, the reflection for WebAssembly is the direct `WebAssembly.Module` object which is being exposed behind a host hook integration. To align this with JS should be to align the `ModuleSource` object and not the `Module` instance object.

guybedford reviewed (changes requested):
Thanks @nicolo-ribaudo for the PR, at a high level the integrations are looking really good here, and the spec work is looking really clean at this point.

The main feedback is around normative changes to the proposal that are also being included in this PR, specifically that the reflection is currently a string in the proposal, and that it directly exposes the `WebAssembly.Module` object instead of making that a getter of an intermediate JS object. This is quite an important property when integrating with the WebAssembly ESM integration to avoid the need to construct intermediate objects that wouldn't otherwise be used.

I'm of course open to further discussion on the above, but as a layering PR feel this work should not be making normative changes to the proposal.

nicolo-ribaudo commented:
Thanks for the review @guybedford! I reverted the normative changes and now I believe that this PR is purely editorial.

I ignored one of your suggestions: instead of tracking the reflection in an `_importReflection_: null | ~module~` variable, I used a boolean. This is because an enum doesn't fit well with the concepts from the hooks refactoring and the other proposals: either you want to execute a module or you want to have it not executed.

nicolo-ribaudo commented:
With these changes, the proposal works without almost any changes to host specs :)

We just need to rename [[WebAssemblyModule]] to [[ModuleSourceObject]] in https://webassembly.github.io/esm-integration/js-api/index.html#esm-integration.

caridy reviewed (approved):
LGTM

guybedford commented on spec.emu:
```diff
@@ -6,77 +6,172 @@
+      <emu-clause id="sec-static-semantics-modulerequests" type="sdo">
+        <h1>
+          Static Semantics: ModuleRequests (
+            <ins>optional _recursiveOnly_: a Boolean</ins>
```
Rather than overloading module requests, perhaps we could create another type of static semantic like `ReflectionRequests`.

nicolo-ribaudo replied:
A `ReflectionRequests` doesn't really work, because for
```js
import module a from "a";
import module b from "b";
import "a";
```

we only want to include `b` in the list of "modules where we should not recurse into", and not `a`. So we always have to do "all the dependencies" − "all the normal dependencies".

If you prefer, I can copy-paste this SDO to create a new `RecursiveModuleRequests` instead of adding the parameter, but keeping them as a single SDO helps keeping them in sync.

guybedford commented on spec.emu:
```diff
@@ -326,6 +598,57 @@ contributors: Luca Casonato, Guy Bedford
+            1. If _body_ is a List of errors, return _body_.
+            1. Let _requestedModules_ be the ModuleRequests of _body_ <ins>with argument *true*</ins>.
+            1. <ins>Let _allRequestedModules_ be the ModuleRequests of _body_.</ins>
+            1. <ins>Let _requestedModuleReflections_ be a new List containing the elements of _allRequestedModules_ that are not in _requestedModules_.</ins>
```
See comment above re a `ReflectionRequests` static semantic here.

guybedford commented on spec.emu:
```diff
@@ -271,6 +463,86 @@ contributors: Luca Casonato, Guy Bedford
+              1. <ins>Let _importOnlyAsReflection_ be *false*.</ins>
+              1. <ins>If _referrer_ has been specified, then</ins>
+                1. <ins>Assert: _specifier_ has been specified.</ins>
+                1. <ins>If _referrer_.[[RequestedModuleReflections]] contains _specifier_, set _importOnlyAsReflection_ to *true*.</ins>
```
Would it be possible to rather have this state passed as another argument to `InnerModuleLoading` instead of needing to check it explicitly here? Also to double-check, `import('specifier')` within the module doesn't hit this path right?

nicolo-ribaudo replied:
Especially now that we have a single list of all the dependencies, we need to keep the check _somehwere_. Doing it here instead of passing it as a parameter to `InnerModuleLoading` avoids the need of passing it as another opaque parameter of HostLoadImportedModule.

guybedford commented on spec.emu:
```diff
@@ -271,6 +463,86 @@ contributors: Luca Casonato, Guy Bedford
+                1. Append _module_ to _state_.[[Visited]].
+                1. Let _requestedModulesCount_ be the length of _module_.[[RequestedModules]].
+                1. Set _state_.[[PendingModulesCount]] to _state_.[[PendingModulesCount]] + _requestedModulesCount_.
+                1. <ins>Let _allRequestedModules_ be the list-concatenation of _module_.[[RequestedModules]] and _module_.[[RequestedModuleReflections]].</ins>
```
Rather than iterating `_allRequestedModules_` we could have two separate iterations, and then set a boolean argument like `reflection` on `InnerModuleLoading`? Alternatively we could even have an `InnerModuleLoadingReflection` to avoid the nested branching.

guybedford reviewed (commented):
Thanks @nicolo-ribaudo, agreed the new changes are completely editorial at this point, and glad to see the simplifications from the layering.

See comments in review.

nicolo-ribaudo commented:
@guybedford I just realized that how we store the two type of imports and how we iterate over them is observable, because both modules evaluation order and modules loading order is observable (loading order is currently only observable by SharedWorkers and Node.js loaders API, but with compartments it would also be observable within ECMA-262).

Consider this example:
```js
import module A from "A";
import module B from "B";
import "C";
import "A";
import "D";
```

With the spec text that I wrote, the loading order is `CADB` and the execution order is `CAD`.

I think that the execution order is fine, but the loading order maybe should be `ABCD`? Wdyt?

caridy commented:
> With the spec text that I wrote, the loading order is `CADB` and the execution order is `CAD`.
> 
> I think that the execution order is fine, but the loading order maybe should be `ABCD`? Wdyt?

I think it should be the one that matches the developer's intuitions, which in this case is `ABCD`, by just looking at the code.

guybedford commented:
It doesn't make a huge difference to users since loads are parallel, but at least from a host consistency perspective, iterating the operations on the imports in pre-order continues to make sense. I think my suggested changes would do that as well?

nicolo-ribaudo commented:
I updated the PR to load dependencies in the correct order.

@guybedford and I discussed about adding a note to discourage hosts from pre-loading dependencies in the `HostLoadImportedModule` hook, since it's also called for import reflections that don't need their dependencies to be loaded. Does this looks good, added as a fourth bullet point to https://ci.tc39.es/preview/tc39/ecma262/sha/24196c3212c3c4ac421e0cf9b002e58fb1445f6d/#sec-HostLoadImportedModule?
```html
<li>
  <p>The host environment should not attempt loading the dependencies of the loaded Module Record.</p>
  <emu-note>
    When using `import module x from "x"`, the dependencies of `"x"` are not immediately needed and may never be needed. If the dependencies of an imported module need to be loaded, it will be done through further calls to HostLoadImportedModule.
  </emu-note>
</li>
```

guybedford commented on spec.emu:
```diff
@@ -326,6 +597,57 @@ contributors: Luca Casonato, Guy Bedford
+          <emu-alg>
+            1. Let _body_ be ParseText(_sourceText_, |Module|).
+            1. If _body_ is a List of errors, return _body_.
+            1. Let _requestedModules_ be the ModuleRequests of _body_.
+            1. <ins>Let _recursiveRequestedModules_ be the ModuleRequests of _body_ with argument *true*</ins>.
+            1. <ins>Let _requestedModuleReflections_ be a list containing the elements of _requestedModules_ that are not elements of _recursiveRequestedModules_
```
To be clearer perhaps let's call this `_requestedModuleReflectionsOnly_` or similarly update the naming / documentation for this field since it is not a list of all of the module reflections now.

nicolo-ribaudo replied:
Done, I also added an `<emu-note>` with an example.

guybedford reviewed (approved):
Great, can confirm this solves the ordering questions. Thanks for talking through the design decisions, let's merge!
